### ISSUE TITLE
Add user configurable check for remaining diskspace.

### ISSUE THREAD

**Issue by manwegit** [borgbackup/borg#721]
I don't know any backup software that does this, but this is a feature that I routinely have to workaround. 

Use case:
Backup target is used by many systems and space usage varies day to day greatly.
Integrity for backup is more important than having finished backup for that particular set.
Currently my automated scripts make disk space checks for backup target before backup starts and report errors / halt backup if 2500 MiB or less space is available. When using remote target this means that I have to make an extra ssh step + checking correct directory for space is PITA. 
Having controlled option in borg to not to do backup if target has less than said value would be nice.

I think limit before backup is enough as everyone needs to adjust the value according what kind of backup they are running, so 2.5 GiB for dedup growth is ok for me.

(If interest for this kind of feature I'll be willing to spend time for a patch + evaluation of said patch)

**Comment by ThomasWaldmann** [comment]
I understand your point, but this is somehow something to be solved outside of borg.

Or how many tools do you know which have builtin free-space checks?

**Comment by manwegit** [comment]
Well that was part of my point, that I don't know backup software that does it. And that missing feature has caused me problems with for example rdiff-backup.

This is a feature that I'd find useful. It might be a feature that others would also appreciate. As such I'd say this is low priority enhancement request.

This missing feature is kind of problem when running backups with authorized_keys lines. You basically don't have any other interface to you backup target at the beginning of backup other than borg. Because remote serve cmd line does not have target folder as a parameter it's not that easy to script with a wrapper.

Another more generic way would be ability to add target path or custom parameters to be passed borg serve. `$SSH_ORIGINAL_COMMAND` would have all options passed that a wrapper script could do what ever with the commands.
Actually I would like this more as it is more generic and leaves diskspace check etc to admin.
(I've done something similar with rsync and modified rrsync script)

The main reason I'd like to have it in borg is that low disk space would not screw up the repo, when there's a minimum buffer requirement for free space (kind of like ext-fs reserved root space).

**Comment by jungle-boogie** [comment]
A noble concern but I'd call this a feature creep.

**Comment by manwegit** [comment]
Adding target to remote serve as cmdline option even if it's not used by borg (so authorized_keys) works, would basically solve all my remote serve problems.

**Comment by ThomasWaldmann** [comment]
Hmm, the (recommended) "borg serve" restriction via authorized_keys is of course in the way if you want to execute something else.

So I guess there are only 2 ways to solve:
a) use generic server monitoring (like nagios/icinga, a cron job that mails you about free space issues or whatever). Maybe you want this for other reasons also, e.g. smartmontools watching your disk health, watching load on the machine, ...
b) we build it into borg, but as @jungle-boogie said, that's a bit feature creep

**Comment by tgharold** [comment]
I use 'monit' for this when I don't want the weight of nagios/ichinga/cactii.

https://mmonit.com/monit/documentation/monit.html

/etc/monit.d/fs-root
```
check filesystem rootfs with path /
  if space usage > 80% then alert
  if inode usage > 80% then alert
```

(I don't remember offhand whether Monit looks at quota restrictions.)

Well worth using a moitoring tool because, as ThomasWaldmann said, you can check other things at the same time (like RAM / swap / CPU usage).  Plus you might have run out of inodes rather then free space (unlikely with 'borg', but common with 'rdiff-backup'), or the server array might be having issues, etc.

There are also bash shell scripts that you can run as a regular user to check free disk and notify yourself if you don't have superuser access.

**Comment by manwegit** [comment]
Well that's the thing. Informing of low space is not enough. We already have monitoring systems and this is secondary location backup from main. I'm also talking about threshold of less than 1%.
I need a way to halt / stop / automatic backup so that the repositories do not get corrupted that I can either redo, or resume when I've made changes to allow more space.  
We have about 1000 repos that backup to number of different mounted targets and not running those that are on the low diskspace directories are the ones that should not backup while others should.

In the past I've used rsync with a wrapper to great benefit and was surprised that target was not an cmdline param for `borg serve`

**Comment by tgharold** [comment]
Makes sense.

What you might want to do is make a SSH command call to the target system and check for the space available on the backup directory prior to running the 'borg create' command.  I have a script (somewhere) that checks for a different issue prior to calling 'borg create', but it could also do a disk space check.

The test for % space free  on the /boot drive would look like:

```test $(df -kP /boot | grep 'boot' | awk '{print $5}' | sed 's/%//g') -gt 5```

I'll try and dig up the wrapper script that I use later.

**Comment by manwegit** [comment]
I already have that kind of setup where I make an extra ssh call and does some preliminary checks. The problem and point is using authorized restrictions on borg call and those checks could then be encapsulated with a wrapper.

**Comment by manwegit** [comment]
This is what I'm proposing: [add location to serve](https://github.com/borgbackup/borg/compare/master...manwegit:feature/remote-param?expand=1)

**Comment by ThomasWaldmann** [comment]
Why are you proposing to "add location to borg serve"?

**Comment by manwegit** [comment]
Wrapper script for borg on remote does not know where borg repo is. That's the place I need to check/verify etc. Having location on serve (even though it's not used internally) would be a generic way to pass that information to remote site allowing more flexible authorized line with wrapper-script.

**Comment by ThomasWaldmann** [comment]
Fixed in master (1.1) by #1336.
